src/compat/db_json.cpp

```diff
@@ -0,0 +1,941 @@
+}
+
+void
+db_json_merge_two_json_objects (JSON_DOC *obj1, JSON_DOC *obj2)
+{
+  assert (db_json_get_type (obj1) == DB_JSON_OBJECT);
+  assert (db_json_get_type (obj2) == DB_JSON_OBJECT);
+
+  for (JSON_VALUE::MemberIterator itr = obj2->MemberBegin ();
+       itr != obj2->MemberEnd (); ++itr)
+    {
+      const char *name = itr->name.GetString ();
+      if (obj1->HasMember (name))
+        {
+          if ((*obj1) [name].IsArray ())
+            {
+              (*obj1) [name].GetArray ().PushBack (itr->value, obj1->GetAllocator ());
+            }
+          else
+            {
+              JSON_VALUE value ((*obj1) [name], obj1->GetAllocator ());
+              (*obj1) [name].SetArray ();
+              (*obj1) [name].PushBack (value, obj1->GetAllocator ());
+              (*obj1) [name].PushBack (itr->value, obj1->GetAllocator ());
+            }
+        }
+      else
+        {
+          obj1->AddMember (itr->name, itr->value, obj1->GetAllocator ());
+        }
+    }
+}
+
+void
+db_json_merge_two_json_arrays (JSON_DOC *array1, JSON_DOC *array2)
+{
+  assert (db_json_get_type (array1) == DB_JSON_ARRAY);
+  assert (db_json_get_type (array2) == DB_JSON_ARRAY);
+
+  for (JSON_VALUE::ValueIterator itr = array2->Begin ();
+       itr != array2->End (); ++itr)
+    {
+      array1->PushBack (*itr, array1->GetAllocator ());
+    }
+}
+
+void
+db_json_merge_two_json_by_array_wrapping (JSON_DOC *j1, JSON_DOC *j2)
+{
+  JSON_DOC *j1_aux = NULL, *j2_aux = NULL, doc;
+
+  if (db_json_get_type (j1) != DB_JSON_ARRAY)
+    {
+      j1_aux = new JSON_DOC;
+      db_json_add_element_to_array (j1_aux, j1);
+    }
+  if (db_json_get_type (j2) != DB_JSON_ARRAY)
+    {
+      j2_aux = new JSON_DOC;
+      db_json_add_element_to_array (j2_aux, j2);
+    }
+
+  db_json_copy_doc (&doc, j1_aux == NULL ? j1 : j1_aux);
+  db_json_merge_two_json_arrays (&doc, j2_aux == NULL ? j2 : j2_aux);
+
+  db_json_copy_doc (j1, &doc);
+  if (j1_aux != NULL)
+    {
+      delete j1_aux;
+    }
+  if (j2_aux != NULL)
+    {
+      delete j2_aux;
+    }
+}
+
+int
+db_json_object_contains_key (JSON_DOC *obj, const char *key, int &error_code)
+{
+  error_code = NO_ERROR;
+
+  if (!obj->IsObject ())
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_NO_JSON_OBJECT_PROVIDED, 0);
+      error_code = ER_NO_JSON_OBJECT_PROVIDED;
+      return 0;
+    }
+
+  return (int) obj->HasMember (key);
+}
+
+char *
+db_json_get_schema_raw_from_validator (JSON_VALIDATOR *val)
+{
+  return val->get_schema_raw();
+}
+
+int
+db_json_validate_json (char *json_body)
+{
+  rapidjson::Document document;
+
+  document.Parse (json_body);
+  if (document.HasParseError ())
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_INVALID_JSON, 2,
+              rapidjson::GetParseError_En (document.GetParseError ()), document.GetErrorOffset ());
+      return ER_INVALID_JSON;
+    }
+
+  return NO_ERROR;
+}
+
+JSON_DOC *db_json_allocate_doc_and_set_type (DB_JSON_TYPE desired_type)
```

Comment:
Seems unnecessary to allocate and set type. Allocate is enough where it is called, while db_json_add_member_to_object and db_json_add_element_to_array can handle null case. They should also handle unacceptable cases (e.g. adding to array on object type json).

Reply:
it isn't unnecessary. when calling AddMember, rapidjson checks whether the provided json doc is an json object or not. if it isn't, the program fails.